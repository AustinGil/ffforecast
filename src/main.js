import URLSearchParams from 'url-search-params';
import { createResponse } from 'create-response';
import { Cookies, SetCookie } from 'cookies';
import createApi from './create-api.js';

/**
 * @param {Date|number|string} date
 * @param {Intl.DateTimeFormatOptions} [options]
 * @returns {string}
 */
export function formatDate(date, options) {
  date = new Date(date);
  return new Intl.DateTimeFormat('en-US', options).format(date);
}

/**
 * @param {number} degree
 */
export function getWindDirection(degree) {
  if (degree >= 337.5 || degree < 22.5) return 'N';
  if (degree < 67.5) return 'NE';
  if (degree < 112.5) return 'E';
  if (degree < 157.5) return 'SE';
  if (degree < 202.5) return 'S';
  if (degree < 247.5) return 'SW';
  if (degree < 292.5) return 'W';
  if (degree < 337.5) return 'NW';
}

/**
 * @param {EW.IngressClientRequest} request
 */
export async function responseProvider(request) {
  const OPENWEATHER_API_KEY = request.getVariable('PMUSER_OPENWEATHER_API_KEY');
  const GEOAPIFY_API_KEY = request.getVariable('PMUSER_GEOAPIFY_API_KEY');
  const api = createApi({
    openWeatherApiKey: OPENWEATHER_API_KEY,
    geoapifyApiKey: GEOAPIFY_API_KEY,
  });
  const headers = {};
  const query = new URLSearchParams(request.query);
  const cookies = new Cookies(request.getHeader('Cookie'));

  let location = cookies.get('location');
  // If user requests new location, update location and cookies
  if (query.has('location')) {
    location = query.get('location');
    headers['Set-Cookie'] = new SetCookie({
      name: 'location',
      value: location,
    }).toHeader();
  }
  // Fallback to server location if no location is set
  if (!location) {
    const { city, region, country } = request.userLocation;
    location = `${city}, ${region}, ${country}`;
  }

  const geolocation = await api.getGeolocation(location);

  let weatherData;
  if (geolocation.features?.length > 0) {
    const { lat, lon } = geolocation.features[0].properties;
    weatherData = await api.oneCall({ lat, lon });
  }

  console.log({ weatherData });

  const body = `<!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="UTF-8" />
      <meta name="viewport" content="width=device-width,initial-scale=1" />
      <title>Freakin' Fast Forecast</title>
      <link rel="icon" href="data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16'><text x='0' y='14'>üå¶</text></svg>">
      <link rel="stylesheet" type="text/css" href="https://unpkg.com/bedrocss">
      <style>
        th, td {
          padding: 0.25rem;
        }
        tr:nth-child(even) {
          background-color: ButtonFace;
        }
        tr:not(:last-child) :is(th, td) {
          border-bottom: 1px solid;
        }
        .relative {
          position: relative;
        }
        .grid {
          display: grid;
        }
        .flex {
          display: flex;
        }
        .flex-grow {
          flex-grow: 1;
        }
        .align-center {
          align-items: center;
        }
        .justify-between {
          justify-content: space-between;
        }
        .min-w-120 {
          min-inline-size: ${120 / 16}rem;
        }
        .max-w-900 {
          max-inline-size: ${900 / 16}rem;
        }
        .o-auto {
          overflow: auto;
        }
        .mi-auto {
          margin-inline: auto;
        }
        ${[0, 4, 8, 12, 16, 32]
          .map((space) => {
            return `
          .gap-${space} {
            gap: ${space / 16}rem;
          }
          .m-${space} {
            margin: ${space / 16}rem;
          }
          .mt-${space} {
            margin-block-start: ${space / 16}rem;
          }
          .mb-${space} {
            margin-block-end: ${space / 16}rem;
          }
          .p-${space} {
            padding: ${space / 16}rem;
          }
          .pb-${space} {
            padding-bottom: ${space / 16}rem;
          }`;
          })
          .join('')}
        .b-1 {
          border: 1px solid;
        }
        .rotate {
          --rotate: 0;
          transform: rotate(var(--rotate));
        }
        .visually-hidden {
          clip: rect(0 0 0 0);
          clip-path: inset(50%);
          height: 1px;
          overflow: hidden;
          position: absolute;
          white-space: nowrap;
          width: 1px;
        }
        .icon {
          display: inline-block;
          width: 1em;
          height: 1em;
          stroke-width: 0;
          stroke: currentColor;
          fill: currentColor;
        }
        .card {
          position: relative;
          border: 1px solid;
          border-radius: 0.5rem;
          padding: .75rem;
          overflow: hidden;
        }
        .card:before {
          content: '';
          position: absolute;
          z-index: -1;
          inset: 0;
          opacity: 0.1;
          background-color: CanvasText;
        }
      </style>
    </head>
  
    <body>
      <main class="max-w-900 mi-auto p-8">
        <h1 class="mb-8">Freakin' Fast Forecast</h1>

        <form class="mb-16">
          <label for="location">Search Location</label>
          <div class="flex gap-8">
            <input id="location" name="location" />
            <button type="submit">Submit</button>
          </div>
        </form>

        ${
          !weatherData
    ? `<p>No weather data for: ${location}</p>`
    : `<h2 class="grid mb-32">Weather for ${location} <br><span class="h6">${formatDate(
      weatherData.current.dt * 1000,
      { dateStyle: 'short', timeStyle: 'short' }
    )}</span></h2>

            <section class="card mb-32">
              <h3 class="h6 mb-8">Current</h3>
              <p class="flex align-center justify-between m-0">
              <span>
                <span class="h1">${weatherData.current.temp}¬∞F</span>
                <br>
                <span>${weatherData.current.weather[0].description}</span>
              </span>
              <img src="http://openweathermap.org/img/wn/${
                weatherData.current.weather[0].icon
    }@2x.png" alt="${weatherData.current.weather[0].description
              }" width="100" height="100"/>
            </p>
            <table role="table" class="mt-0">
              <tbody role="rowgroup">
                <tr role="row">
                  <th scope="row" role="rowheader" align="left"><span aria-hidden="true">üå°</span> Feels Like</th>
                  <td role="cell">${weatherData.current.feels_like}¬∞F</td>
                </tr>
                <tr role="row">
                  <th scope="row" role="rowheader" align="left"><span aria-hidden="true">üå¨</span> Wind</th>
                  <td role="cell">
                  ${weatherData.current.wind_speed} mph
                  
                  <svg class="icon icon-arrow-long-up rotate" style="--rotate: ${
    weatherData.current.wind_deg
                  }deg;" aria-hidden="true" alt=""><use xlink:href="#icon-arrow-long-up"></use></svg>

                    ${getWindDirection(weatherData.current.wind_deg)}
                  </td>
                </tr>
                <tr role="row">
                  <th scope="row" role="rowheader" align="left"><span aria-hidden="true">‚òÅ</span> Cloudiness</th>
                  <td role="cell">${weatherData.current.clouds}%</td>
                </tr>
                <tr role="row">
                  <th scope="row" role="rowheader" align="left"><span aria-hidden="true">üï∂</span> UV Index</th>
                  <td role="cell">${weatherData.current.uvi}</td>
                </tr>
              </tbody>
            </table>
          </section>

          <section class="card mb-32">
            <h3 class="h6 mb-8">Hourly</h3>
            <div class="flex gap-8 o-auto pb-8">
              ${weatherData.hourly
      .map((hour, index) => {
        return `
              <div class="min-w-120">
                <small>
                ${formatDate(hour.dt * 1000, {
                  dateStyle: 'short',
                })}
                </small>
                <br>
                <h4><b>${formatDate(hour.dt * 1000, {
                  timeStyle: 'short',
                })}</b></h4>
                
                <p class="h5 flex justify-between align-center mt-0">
                  <b>${Math.round(hour.temp)}¬∞F</b>
                  <img
                    src="http://openweathermap.org/img/wn/${
                    hour.weather[0].icon
                    }@2x.png"
                    alt="${hour.weather[0].description}"
                    loading="${index < 3 ? 'eager' : 'lazy'}"
                    width="64"
                    height="64"
                  />
                </p>

                <table role="table" class="mt-0">
                  <tbody role="rowgroup">
                    <tr role="row">
                      <th scope="row" role="rowheader" align="left">
                        <span aria-hidden="true" title="Cloudiness">‚òÅ</span>
                        <span class="visually-hidden">Cloudiness</span>
                      </th>
                      <td role="cell">${hour.clouds}%</td>
                    </tr>
                    <tr role="row">
                      <th scope="row" role="rowheader" align="left">
                        <span aria-hidden="true" title="Precipitation">üåß</span>
                        <span class="visually-hidden">Precipitation</span>
                      </th>
                      <td role="cell">${hour.pop}%</td>
                    </tr>
                    <tr role="row">
                      <th scope="row" role="rowheader" align="left">
                        <span aria-hidden="true" title="Humidity">ü•µ</span>
                        <span class="visually-hidden">Humidity</span>
                      </th>
                      <td role="cell">${hour.humidity}%</td>
                    </tr>
                    <tr role="row">
                      <th scope="row" role="rowheader" align="left">
                        <span aria-hidden="true" title="UV Index">üï∂</span>
                        <span class="visually-hidden">UV Index</span>
                      </th>
                      <td role="cell">${hour.uvi}%</td>
                    </tr>
                  </tbody>
                </table>
              </div>`;
                })
      .join('')}
            </div>
          </section>

          <section class="card mb-32">
            <h3 class="h6 mb-8">Daily</h3>
            <div class="flex gap-8 o-auto pb-8">
              ${weatherData.daily
      .map((day, index) => {
                  return `<div class="flex-grow min-w-120">
                          <h4><b>${formatDate(day.dt * 1000, {
                            dateStyle: 'short',
                          })}</b></h4>

                          <p class="h5 flex justify-between align-center mt-0">
                            <b>${Math.round(day.temp.day)}¬∞F</b>
                            <img
                              src="http://openweathermap.org/img/wn/${
                    day.weather[0].icon
                              }@2x.png"
                              alt="${day.weather[0].description}"
                              loading="${index < 3 ? 'eager' : 'lazy'}"
                              width="64"
                              height="64"
                            />
                          </p>
                          <table role="table" class="mt-0">
                            <tbody role="rowgroup">
                              <tr role="row">
                                <th scope="row" role="rowheader" align="left">
                                  <span aria-hidden="true" title="High">üî∫</span>
                                  <span class="visually-hidden">High</span>
                                </th>
                                <td role="cell">${day.temp.max}¬∞F</td>
                              </tr>
                              <tr role="row">
                                <th scope="row" role="rowheader" align="left">
                                  <span aria-hidden="true" title="Low">üîª</span>
                                  <span class="visually-hidden">Low</span>
                                </th>
                                <td role="cell">${day.temp.min}¬∞F</td>
                              </tr>
                              <tr role="row">
                                <th scope="row" role="rowheader" align="left">
                                  <span aria-hidden="true" title="Cloudiness">‚òÅ</span>
                                  <span class="visually-hidden">Cloudiness</span>
                                </th>
                                <td role="cell">${day.clouds}%</td>
                              </tr>
                              <tr role="row">
                                <th scope="row" role="rowheader" align="left">
                                  <span aria-hidden="true" title="Precipitation">üåß</span>
                                  <span class="visually-hidden">Precipitation</span>
                                </th>
                                <td role="cell">${day.pop}%</td>
                              </tr>
                              <tr role="row">
                                <th scope="row" role="rowheader" align="left">
                                  <span aria-hidden="true" title="UV Index">üï∂</span>
                                  <span class="visually-hidden">UV Index</span>
                                </th>
                                <td role="cell">${day.uvi}¬∞F</td>
                              </tr>
                            </tbody>
                          </table>
                        </div>`;
                })
      .join('')}
            </div>
            `
        }
        </section>
      </main>
      <svg aria-hidden="true" style="position: absolute; width: 0; height: 0; overflow: hidden;" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
      <defs>
      <symbol id="icon-arrow-long-up" viewBox="0 0 20 20">
      <path d="M10 0.75l5.5 5.25h-3.5v13h-4v-13h-3.5l5.5-5.25z"></path>
      </symbol>
      </defs>
      </svg>
    </body>
  </html>`;

  return createResponse(body, {
    status: 200,
    headers: headers,
  });
}
