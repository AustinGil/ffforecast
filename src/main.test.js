import { vi, describe, test, expect } from 'vitest';
import { faker } from '@faker-js/faker';
import fetch from 'node-fetch';
import { formatDate, getWindDirection, responseProvider } from './main.js';

vi.mock('node-fetch', () => {
  return {
    default: vi.fn().mockResolvedValue({
      json: vi.fn().mockResolvedValue({
        favoriteAnimal: faker.animal.dog(),
      }),
    }),
  };
});

describe('formatDate', () => {
  const randomDate = faker.date.recent();

  test.concurrent('it returns a date in M/D/YYYY', () => {
    const formatted = formatDate(randomDate);
    expect(formatted).toBe(randomDate.toLocaleDateString());
  });
  test.concurrent('it accepts a Date, number, or string', () => {
    const dateDate = new Date(randomDate);
    expect(formatDate(dateDate)).toBe(randomDate.toLocaleDateString());
    const numberDate = Number(randomDate);
    expect(formatDate(numberDate)).toBe(randomDate.toLocaleDateString());
    const stringDate = String(randomDate);
    expect(formatDate(stringDate)).toBe(randomDate.toLocaleDateString());
  });
  test.concurrent('it works with Intl.DateTimeFormatOptions', () => {
    /** @type {Intl.DateTimeFormatOptions} */
    const options = {
      dateStyle: 'short',
      timeStyle: 'short',
    };
    const formatted = formatDate(randomDate, options);
    expect(formatted).toBe(
      new Intl.DateTimeFormat('en-US', options).format(randomDate)
    );
  });
});

describe('getWindDirection', () => {
  test.concurrent('returns NE for North-East degree range', () => {
    expect(getWindDirection(23)).toBe('NE');
    expect(getWindDirection(67)).toBe('NE');
  });
  test.concurrent('returns E for East degree range', () => {
    expect(getWindDirection(68)).toBe('E');
    expect(getWindDirection(112)).toBe('E');
  });
  test.concurrent('returns SE for South-East degree range', () => {
    expect(getWindDirection(113)).toBe('SE');
    expect(getWindDirection(157)).toBe('SE');
  });
  test.concurrent('returns S for South degree range', () => {
    expect(getWindDirection(158)).toBe('S');
    expect(getWindDirection(202)).toBe('S');
  });
  test.concurrent('returns SW for South-West degree range', () => {
    expect(getWindDirection(203)).toBe('SW');
    expect(getWindDirection(247)).toBe('SW');
  });
  test.concurrent('returns W for West degree range', () => {
    expect(getWindDirection(248)).toBe('W');
    expect(getWindDirection(292)).toBe('W');
  });
  test.concurrent('returns NW for North-West degree range', () => {
    expect(getWindDirection(293)).toBe('NW');
    expect(getWindDirection(337)).toBe('NW');
  });
  test.concurrent('returns N for North degree range', () => {
    expect(getWindDirection(338)).toBe('N');
    expect(getWindDirection(22)).toBe('N');
  });
});

describe('responseProvider', () => {
  /** @type {EW.IngressClientRequest} */
  // @ts-ignore
  const fakeRequestObject = {
    getVariable: vi.fn((key) => key),
    query: undefined,
    getHeader: vi.fn(),
    userLocation: {
      continent: undefined,
      country: faker.address.country(),
      region: faker.address.state(),
      city: faker.address.city(),
      zipCode: faker.address.zipCode(),
    },
  };

  test.concurrent('returns a valid response object', async () => {
    const response = await responseProvider(fakeRequestObject);
    expect(typeof response.body).toBe('string');
    expect(response).toMatchObject({
      status: 200,
      headers: {},
      deny_reason: '',
    });
  });
  test.concurrent('default UI matches snapshot', async () => {
    const response = await responseProvider({
      ...fakeRequestObject,
      userLocation: {
        country: 'Gameboydia',
        region: 'Hoenn',
        city: 'Pallet Town',
        zipCode: '12345',
        continent: 'Kanto',
      },
    });
    expect(response).toMatchSnapshot();
  });
});
