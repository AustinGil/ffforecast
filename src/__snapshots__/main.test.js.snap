// Vitest Snapshot v1

exports[`responseProvider > default UI matches snapshot 1`] = `
{
  "body": "<!DOCTYPE html>
  <html lang=\\"en\\">
    <head>
      <meta charset=\\"UTF-8\\" />
      <meta name=\\"viewport\\" content=\\"width=device-width,initial-scale=1\\" />
      <title>Freakin' Fast Forecast</title>
      <link rel=\\"icon\\" href=\\"data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16'><text x='0' y='14'>ðŸŒ¦</text></svg>\\">
      <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"https://unpkg.com/bedrocss\\">
      <style>
        th, td {
          padding: 0.25rem;
        }
        tr:nth-child(even) {
          background-color: ButtonFace;
        }
        tr:not(:last-child) :is(th, td) {
          border-bottom: 1px solid;
        }
        .relative {
          position: relative;
        }
        .grid {
          display: grid;
        }
        .flex {
          display: flex;
        }
        .flex-grow {
          flex-grow: 1;
        }
        .align-center {
          align-items: center;
        }
        .justify-between {
          justify-content: space-between;
        }
        .min-w-120 {
          min-inline-size: 7.5rem;
        }
        .max-w-900 {
          max-inline-size: 56.25rem;
        }
        .o-auto {
          overflow: auto;
        }
        .mi-auto {
          margin-inline: auto;
        }
        
          .gap-0 {
            gap: 0rem;
          }
          .m-0 {
            margin: 0rem;
          }
          .mt-0 {
            margin-block-start: 0rem;
          }
          .mb-0 {
            margin-block-end: 0rem;
          }
          .p-0 {
            padding: 0rem;
          }
          .pb-0 {
            padding-bottom: 0rem;
          }
          .gap-4 {
            gap: 0.25rem;
          }
          .m-4 {
            margin: 0.25rem;
          }
          .mt-4 {
            margin-block-start: 0.25rem;
          }
          .mb-4 {
            margin-block-end: 0.25rem;
          }
          .p-4 {
            padding: 0.25rem;
          }
          .pb-4 {
            padding-bottom: 0.25rem;
          }
          .gap-8 {
            gap: 0.5rem;
          }
          .m-8 {
            margin: 0.5rem;
          }
          .mt-8 {
            margin-block-start: 0.5rem;
          }
          .mb-8 {
            margin-block-end: 0.5rem;
          }
          .p-8 {
            padding: 0.5rem;
          }
          .pb-8 {
            padding-bottom: 0.5rem;
          }
          .gap-12 {
            gap: 0.75rem;
          }
          .m-12 {
            margin: 0.75rem;
          }
          .mt-12 {
            margin-block-start: 0.75rem;
          }
          .mb-12 {
            margin-block-end: 0.75rem;
          }
          .p-12 {
            padding: 0.75rem;
          }
          .pb-12 {
            padding-bottom: 0.75rem;
          }
          .gap-16 {
            gap: 1rem;
          }
          .m-16 {
            margin: 1rem;
          }
          .mt-16 {
            margin-block-start: 1rem;
          }
          .mb-16 {
            margin-block-end: 1rem;
          }
          .p-16 {
            padding: 1rem;
          }
          .pb-16 {
            padding-bottom: 1rem;
          }
          .gap-32 {
            gap: 2rem;
          }
          .m-32 {
            margin: 2rem;
          }
          .mt-32 {
            margin-block-start: 2rem;
          }
          .mb-32 {
            margin-block-end: 2rem;
          }
          .p-32 {
            padding: 2rem;
          }
          .pb-32 {
            padding-bottom: 2rem;
          }
        .b-1 {
          border: 1px solid;
        }
        .rotate {
          --rotate: 0;
          transform: rotate(var(--rotate));
        }
        .visually-hidden {
          clip: rect(0 0 0 0);
          clip-path: inset(50%);
          height: 1px;
          overflow: hidden;
          position: absolute;
          white-space: nowrap;
          width: 1px;
        }
        .icon {
          display: inline-block;
          width: 1em;
          height: 1em;
          stroke-width: 0;
          stroke: currentColor;
          fill: currentColor;
        }
        .card {
          position: relative;
          border: 1px solid;
          border-radius: 0.5rem;
          padding: .75rem;
          overflow: hidden;
        }
        .card:before {
          content: '';
          position: absolute;
          z-index: -1;
          inset: 0;
          opacity: 0.1;
          background-color: CanvasText;
        }
      </style>
    </head>
  
    <body>
      <main class=\\"max-w-900 mi-auto p-8\\">
        <h1 class=\\"mb-8\\">Freakin' Fast Forecast</h1>

        <form class=\\"mb-16\\">
          <label for=\\"location\\">Search Location</label>
          <div class=\\"flex gap-8\\">
            <input id=\\"location\\" name=\\"location\\" />
            <button type=\\"submit\\">Submit</button>
          </div>
        </form>

        <p>No weather data for: Pallet Town, Hoenn, Gameboydia</p>
        </section>
      </main>
      <svg aria-hidden=\\"true\\" style=\\"position: absolute; width: 0; height: 0; overflow: hidden;\\" version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\">
      <defs>
      <symbol id=\\"icon-arrow-long-up\\" viewBox=\\"0 0 20 20\\">
      <path d=\\"M10 0.75l5.5 5.25h-3.5v13h-4v-13h-3.5l5.5-5.25z\\"></path>
      </symbol>
      </defs>
      </svg>
    </body>
  </html>",
  "deny_reason": "",
  "headers": {},
  "status": 200,
}
`;
